core/src/main/java/hudson/model/AbstractCIBase.java

```diff
@@ -89,17 +89,29 @@ protected void killComputer(Computer c) {
         c.kill();
     }
 
-    /* =================================================================================================================
-    * Package-protected, but accessed API
-    * ============================================================================================================== */
+    private final Set<String> disabledAdministrativeMonitors = Collections.synchronizedSet(new HashSet<>());
```

Comment:
The docs of synchronizedSet says (I've never used this before)

```
Returns a synchronized (thread-safe) set backed by the specified set. In order to guarantee serial access, it is critical that all access to the backing set is accomplished through the returned set.

It is imperative that the user manually synchronize on the returned set when iterating over it:

  Set s = Collections.synchronizedSet(new HashSet());
      ...
  synchronized (s) {
      Iterator i = s.iterator(); // Must be in the synchronized block
      while (i.hasNext())
          foo(i.next());
  }
 

Failure to follow this advice may result in non-deterministic behavior. 
```
I'm not sure what it means by ` Failure to follow this advice may result in non-deterministic behavior. ` in here

Reply:
There's no direct iteration of this field,

there's 4 usages of it,

* enable / disable, calls add /remove
* isEnabled, calls `contains`
* the UI, which iterates over it, not sure if there's much we can do there,
* jcasc, it will use the setter to update it which will clear the set and use what jcasc sets.

Reply:
My worry is what is the worst that could happen? Will it throw an exception or will it just show old data?
If its the latter then I think its alright but the documentation doesn't seem perfectly clear

Reply:
https://github.com/jenkinsci/jenkins/pull/3687#issuecomment-430732957 was the (un)funnest one I've seen. Way up there in my hall of shame ☹️

Reply:
That's only relevant because iterating over a collection is technically _n_ operations, not 1. You need to synchronize for the duration of the logical operation, not _n_ separate synchronizations for _n_ items in the collection. This is all about fencing the allowed execution timeline based on "happens-before" rules in Java.

So, in a sense, no, peppering `synchronized` and `volatile` everywhere isn't good enough to solve every concurrency issue! 😅